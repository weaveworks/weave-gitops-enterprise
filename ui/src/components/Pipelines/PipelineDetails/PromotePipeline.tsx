import React from 'react';
import ShowChartIcon from '@material-ui/icons/ShowChart';
import { Button, Flex, Link, Text } from '@weaveworks/weave-gitops';
import { ApprovePromotionRequest } from '../../../api/pipelines/pipelines.pb';

import { useApprove } from '../../../hooks/pipelines';

const PromotePipeline = ({
  req,
  promoteVersion,
}: {
  req: ApprovePromotionRequest;
  promoteVersion: string;
}) => {
  const { data, isLoading, refetch } = useApprove(req);

  return (
    <Flex column gap="4">
      <Button
        startIcon={<ShowChartIcon />}
        onClick={() => refetch()}
        disabled={isLoading || !promoteVersion}
        loading={isLoading}
      >
        Approve Promotion
      </Button>
      {/* <Text color="primary20">
        PR:
        {data ? (
          <Link href={data.pullRequestURL} newTab>
            {data.pullRequestURL}
          </Link>
        ) : (
          ' Waiting For Approval'
        )}
      </Text> */}
    </Flex>
  );
};

export default PromotePipeline;
