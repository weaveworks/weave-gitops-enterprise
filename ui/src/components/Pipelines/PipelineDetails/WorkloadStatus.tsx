import {
  Flex,
  Icon,
  IconType,
  KubeStatusIndicator,
} from '@weaveworks/weave-gitops';
import styled from 'styled-components';
import { WorkloadStatus as WorkloadStatusType } from '../../../api/pipelines/types.pb';

interface Props {
  className?: string;
  workload?: WorkloadStatusType;
}

function WorkloadStatus({ className, workload }: Props) {
  return (
    <div className={className}>
      <Flex align>
        <div style={{ marginRight: 4 }}>
          {workload?.suspended ? (
            <Icon
              size="small"
              type={IconType.SuspendedIcon}
              color="suspended"
            />
          ) : (
            <KubeStatusIndicator
              short
              conditions={workload?.conditions || []}
            />
          )}
        </div>
        <div>{workload?.name}</div>
      </Flex>
    </div>
  );
}

export default styled(WorkloadStatus).attrs({ className: 'WorkloadStatus' })`
  font-size: ${props => props.theme.fontSizes.large};

  /* Hack to hide the text; we only want the icon here */
  ${KubeStatusIndicator} span {
    display: none;
  }

  ${Icon} img {
    width: 16px;
  }
`;
