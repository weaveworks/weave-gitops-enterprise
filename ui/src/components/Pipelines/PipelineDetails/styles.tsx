import { createStyles, makeStyles } from '@material-ui/core';
import styled from 'styled-components';

export const usePipelineStyles = makeStyles(() =>
  createStyles({
    gridContainer: {
      margin: `12px 12px`,
      padding: '24px',
      borderRadius: '10px',
    },
    gridWrapper: {
      width: '100%',
      display: 'flex',
      flexWrap: 'nowrap',
      overflow: 'auto',
      paddingBottom: '8px',
      margin: `24 0 0 0`,
    },
    subtitleColor: {
      color: '#737373',
    },
    editButton: {
      paddingBottom: '12px',
    },
  }),
);
export const TargetWrapper = styled.div`
  font-size: ${props => props.theme.fontSizes.large};
  margin-bottom: ${props => props.theme.spacing.small};
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  width: calc(250px - ${props => props.theme.spacing.large});
`;
export const CardContainer = styled.div`
  background: ${props => props.theme.colors.white};
  padding: ${props => props.theme.spacing.small};
  margin-bottom: ${props => props.theme.spacing.xs};
  box-shadow: 0px 0px 1px rgba(26, 32, 36, 0.32);
  border-radius: 10px;
  font-weight: 600;
`;
export const Title = styled.div`
  font-size: ${props => props.theme.fontSizes.medium};
  color: ${props => props.theme.colors.black};
  font-weight: 400;
`;
export const ClusterName = styled.div`
  margin-bottom: ${props => props.theme.spacing.small};
  line-height: 24px;
  a > span {
    font-size: 20px;
  }
`;
export const TargetNamespace = styled.div`
  font-size: ${props => props.theme.fontSizes.medium};
`;
export const WorkloadWrapper = styled.div`
  position: relative;
  .version {
    margin-left: ${props => props.theme.spacing.xxs};
  }
`;
export const LastAppliedVersion = styled.span`
  color: ${props => props.theme.colors.neutral30};
  font-size: ${props => props.theme.fontSizes.medium};
  border: 1px solid ${props => props.theme.colors.neutral20};
  padding: 14px 6px;
  border-radius: 50%;
`;
