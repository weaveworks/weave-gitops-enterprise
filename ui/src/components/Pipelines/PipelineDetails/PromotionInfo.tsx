import { Flex } from '@weaveworks/weave-gitops';
import _ from 'lodash';
import styled from 'styled-components';
import { PipelineTargetStatus } from '../../../api/pipelines/types.pb';
import { EnvironmentCard } from './styles';

type Props = {
  className?: string;
  targets: PipelineTargetStatus[];
};

const PromotionInfoContainer = styled(EnvironmentCard)`
  background: #f6f7f9;
  border: 1px solid black;
  border-style: dashed;
`;
export enum ReadyType {
  Ready = 'Ready',
  NotReady = 'Not Ready',
  Reconciling = 'Reconciling',
  PendingAction = 'PendingAction',
  Suspended = 'Suspended',
  None = 'None',
}

export enum ReadyStatusValue {
  True = 'True',
  False = 'False',
  Unknown = 'Unknown',
  None = 'None',
}

export function computeReady(conditions: any): ReadyType | undefined {
  if (!conditions?.length) return undefined;
  const readyCondition = _.find(
    conditions,
    c => c.type === 'Ready' || c.type === 'Available',
  );

  if (readyCondition) {
    if (readyCondition.status === ReadyStatusValue.True) {
      return ReadyType.Ready;
    }

    if (readyCondition.status === ReadyStatusValue.Unknown) {
      if (readyCondition.reason === 'Progressing') return ReadyType.Reconciling;
      if (readyCondition.reason === 'TerraformPlannedWithChanges')
        return ReadyType.PendingAction;
    }

    if (readyCondition.status === ReadyStatusValue.None) return ReadyType.None;

    return ReadyType.NotReady;
  }

  if (_.find(conditions, c => c.status === ReadyStatusValue.False)) {
    return ReadyType.NotReady;
  }

  return ReadyType.Ready;
}
const gatherTargetStatus = (targets: PipelineTargetStatus[]) => {
  const reduced = targets.reduce(
    (obj, target) => {
      let ready = true;
      target.workloads?.forEach(workload => {
        if (computeReady(workload.conditions || []) !== 'Ready') ready = false;
      });
      if (ready) obj.true += 1;
      else obj.false += 1;
      return obj;
    },
    { true: 0, false: 0 },
  );
  return `${reduced['true']} out of ${targets.length} targets ready for promotion`;
};

function PromotionInfo({ className, targets }: Props) {
  return (
    <PromotionInfoContainer className={className}>
      <Flex center align wide tall wrap>
        {gatherTargetStatus(targets)}
      </Flex>
    </PromotionInfoContainer>
  );
}

export default styled(PromotionInfo).attrs({ className: PromotionInfo.name })``;
