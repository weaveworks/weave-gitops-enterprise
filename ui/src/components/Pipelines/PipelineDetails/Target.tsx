import {
  Flex,
  Icon,
  IconType,
  KubeStatusIndicator,
  Link,
  Text,
  V2Routes,
  formatURL,
} from '@weaveworks/weave-gitops';
import _ from 'lodash';
import styled from 'styled-components';
import { PipelineTargetStatus } from '../../../api/pipelines/types.pb';
import { useListConfigContext } from '../../../contexts/ListConfig';
import { ClusterDashboardLink } from '../../Clusters/ClusterDashboardLink';
import { EnvironmentCard } from './styles';

type Props = {
  className?: string;
  target: PipelineTargetStatus;
  background: number;
};

function Target({ className, target, background }: Props) {
  const configResponse = useListConfigContext();
  const clusterName = target?.clusterRef?.name
    ? `${target?.clusterRef?.namespace || 'default'}/${
        target?.clusterRef?.name
      }`
    : configResponse?.data?.managementClusterName || 'undefined';

  //questions for Monday: clusterRefs empty? Cluster namespace? Multiple workloads?
  return (
    <EnvironmentCard
      className={className}
      background={background}
      column
      gap="8"
    >
      <Flex column>
        <Flex gap="4" align>
          <Icon type={IconType.ClustersIcon} size="medium" color="neutral20" />
          <Text>Cluster:</Text>
        </Flex>
        <ClusterDashboardLink clusterName={clusterName} />
      </Flex>
      {_.map(target.workloads, (workload, index) => {
        return (
          <Flex key={index} column gap="8">
            <Flex column>
              <Text size="medium">Namespace/Name</Text>
              <Flex gap="4" align>
                <KubeStatusIndicator
                  noText
                  conditions={workload?.conditions || []}
                />
                <Link
                  to={formatURL(V2Routes.HelmRelease, {
                    name: workload.name,
                    namespace: target.namespace,
                    clusterName: clusterName,
                  })}
                >
                  {target.namespace} / {workload.name}
                </Link>
              </Flex>
            </Flex>
            <Text bold size="small">
              LAST APPLIED VERSION: {'V' + workload.lastAppliedRevision || '-'}
            </Text>
            <Text size="small">
              SPECIFIED VERSION: {'V' + workload.version || '-'}
            </Text>
          </Flex>
        );
      })}
    </EnvironmentCard>
  );
}

export default styled(Target).attrs({ className: Target.name })``;
