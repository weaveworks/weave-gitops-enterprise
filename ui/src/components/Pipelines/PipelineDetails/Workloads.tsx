import { Flex, Text } from '@weaveworks/weave-gitops';
import _ from 'lodash';
import styled from 'styled-components';
import { Pipeline, Promotion } from '../../../api/pipelines/types.pb';
import PromotePipeline from './PromotePipeline';
import PromotionInfo from './PromotionInfo';
import Target from './Target';
import { EnvironmentCard } from './styles';

const getStrategy = (promo?: Promotion) => {
  if (!promo) return '-';
  if (!promo.manual) return 'Automated';

  const nonNullStrat = _.map(promo.strategy, (value, key) => {
    if (value !== null) return key;
  });
  return _.startCase(nonNullStrat[0] || '-');
};

const EnvironmentContainer = styled(Flex)`
  background: ${props => props.theme.colors.pipelineGray}};
  border-radius: 8px;
  padding: ${props => props.theme.spacing.small};
`;

const PromotionContainer = styled.div`
  height: 40px;
  padding: ${props => props.theme.spacing.small} 0;
`;

function Workloads({
  pipeline,
  className,
}: {
  pipeline: Pipeline;
  className?: string;
}) {
  const environments = pipeline?.environments || [];
  const targetsStatuses = pipeline?.status?.environments || {};

  return (
    <Flex gap="4" tall wide className={className}>
      {environments.map((env, index) => {
        const status = targetsStatuses[env.name!].targetsStatuses || [];
        const promoteVersion =
          targetsStatuses[env.name!].waitingStatus?.revision || '';
        const strategy = env.promotion
          ? getStrategy(env.promotion)
          : getStrategy(pipeline.promotion);

        return (
          <EnvironmentContainer key={index} tall column gap="16">
            <PromotionInfo targets={status} />
            <EnvironmentCard background={index}>
              <Flex column gap="8" wide>
                <Flex between align wide>
                  <Text bold capitalize size="large">
                    {env.name}
                  </Text>
                  <Text>{env.targets?.length || '0'} TARGETS</Text>
                </Flex>
                <Flex gap="8" wide start>
                  <Text bold>Strategy:</Text>
                  <Text> {strategy}</Text>
                </Flex>
              </Flex>
              <PromotionContainer>
                {strategy !== 'Automated' &&
                  index < environments.length - 1 && (
                    <PromotePipeline
                      req={{
                        name: pipeline.name,
                        env: env.name,
                        namespace: pipeline.namespace,
                        revision: promoteVersion,
                      }}
                      promoteVersion={promoteVersion || ''}
                    />
                  )}
              </PromotionContainer>
            </EnvironmentCard>
            {status.map((target, indx) => (
              <Target key={indx} target={target} background={index} />
            ))}
          </EnvironmentContainer>
        );
      })}
    </Flex>
  );
}

export default styled(Workloads)`
  background: ${props => props.theme.colors.backGray}};
  padding: ${props => props.theme.spacing.medium};
  overflow-x: auto;
  box-sizing: border-box;
`;
