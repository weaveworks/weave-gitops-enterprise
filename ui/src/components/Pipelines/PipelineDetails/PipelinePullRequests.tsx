import { DataTable, Flex, Link } from '@weaveworks/weave-gitops';
import _ from 'lodash';
import styled from 'styled-components';
import { Pipeline } from '../../../api/pipelines/types.pb';
import { useGetPullRequestsForPipeline } from '../../../contexts/Pipelines';

type Props = {
  className?: string;
  pipeline?: Pipeline;
};

function PipelinePullRequests({ className, pipeline }: Props) {
  const { data: prs } = useGetPullRequestsForPipeline(pipeline);

  const rows = _.map(prs?.pullRequests, (url, env) => ({ env, url }));

  return (
    <div className={className}>
      <Flex wide>
        <DataTable
          fields={[
            { label: 'Environment', value: 'env' },
            {
              label: 'URL',
              value: r => (
                <Link href={r.url} newTab>
                  {r.url}
                </Link>
              ),
            },
          ]}
          rows={rows}
        />
      </Flex>
    </div>
  );
}

export default styled(PipelinePullRequests).attrs({
  className: PipelinePullRequests.name,
})`
  width: 100%;
`;
