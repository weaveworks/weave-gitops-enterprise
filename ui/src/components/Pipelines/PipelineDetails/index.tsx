import { Box } from '@material-ui/core';
import { ListError } from '@weaveworks/progressive-delivery/api/prog/types.pb';
import {
  Flex,
  RouterTab,
  SubRouterTabs,
  YamlView,
} from '@weaveworks/weave-gitops';
import { GetPipelineResponse } from '../../../api/pipelines/pipelines.pb';
import { Pipeline } from '../../../api/pipelines/types.pb';
import { useGetPipeline } from '../../../contexts/Pipelines';
import { Routes } from '../../../utils/nav';
import KeyValueTable, { KeyValuePairs } from '../../KeyValueTable';
import { Page } from '../../Layout/App';
import { NotificationsWrapper } from '../../Layout/NotificationsWrapper';
import { EditButton } from './../../../components/Templates/Edit/EditButton';
import PipelinePullRequests from './PipelinePullRequests';
import Workloads from './Workloads';
import { usePipelineStyles } from './styles';

const mappedErrors = (
  errors: Array<string>,
  namespace: string,
): Array<ListError> => {
  return errors.map(err => ({
    message: err,
    namespace,
  }));
};
interface Props {
  name: string;
  namespace: string;
}

const pipelineStrategyText = (data?: GetPipelineResponse): KeyValuePairs => {
  const pairs: KeyValuePairs = [];

  if (!data) {
    return pairs;
  }

  // Trying to differentiate between null and negative values
  if (data?.pipeline?.promotion?.manual === null) {
    pairs.push(['Automated', null]);
  } else {
    pairs.push([
      'Automated',
      data?.pipeline?.promotion?.manual ? 'False' : 'True',
    ]);
  }

  const strat = data?.pipeline?.promotion?.strategy;

  if (strat?.pullRequest === null && strat?.notification === null) {
    pairs.push(['Strategy', null]);
  } else {
    pairs.push([
      'Strategy',
      data?.pipeline?.promotion?.strategy?.pullRequest
        ? 'Pull Request'
        : 'Notification',
    ]);

    if (strat?.pullRequest !== null) {
      const pr = data.pipeline?.promotion?.strategy?.pullRequest;
      pairs.push(['URL', pr?.url]);
      pairs.push(['Branch', pr?.branch]);
    }
  }

  return pairs;
};

const PipelineDetails = ({ name, namespace }: Props) => {
  const { isLoading, data } = useGetPipeline({
    name,
    namespace,
  });

  const classes = usePipelineStyles();
  const path = `/pipelines/details`;

  return (
    <Page
      loading={isLoading}
      path={[
        {
          label: 'Pipelines',
          url: Routes.Pipelines,
        },
        {
          label: name,
        },
      ]}
    >
      <NotificationsWrapper
        errors={mappedErrors(data?.errors || [], namespace)}
      >
        <Box marginBottom={2}>
          <Flex align wide between>
            <KeyValueTable pairs={pipelineStrategyText(data)} />
            <div>
              <EditButton
                className={classes.editButton}
                resource={data?.pipeline || ({} as Pipeline)}
              />
            </div>
          </Flex>
        </Box>

        <SubRouterTabs rootPath={`${path}/status`}>
          <RouterTab name="Status" path={`${path}/status`}>
            <Workloads pipeline={data?.pipeline || ({} as Pipeline)} />
          </RouterTab>
          <RouterTab name="Yaml" path={`${path}/yaml`}>
            <YamlView
              yaml={data?.pipeline?.yaml || ''}
              object={{
                kind: 'Pipeline',
                name: data?.pipeline?.name,
                namespace: data?.pipeline?.namespace,
              }}
            />
          </RouterTab>
          <RouterTab name="Pull Requests" path={`${path}/pullrequests`}>
            <PipelinePullRequests pipeline={data?.pipeline} />
          </RouterTab>
        </SubRouterTabs>
      </NotificationsWrapper>
    </Page>
  );
};

export default PipelineDetails;
