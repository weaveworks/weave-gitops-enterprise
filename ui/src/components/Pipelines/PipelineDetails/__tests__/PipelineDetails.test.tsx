/* eslint-disable testing-library/no-node-access */
import { act, render, screen } from '@testing-library/react';
import { CoreClientContextProvider } from '@weaveworks/weave-gitops';
import PipelineDetails from '..';
import { GetPipelineResponse } from '../../../../api/pipelines/pipelines.pb';
import { Pipeline } from '../../../../api/pipelines/types.pb';
import { EnterpriseClientContext } from '../../../../contexts/API';
import {
  CoreClientMock,
  PipelinesClientMock,
  defaultContexts,
  withContext,
} from '../../../../utils/test-utils';

const res: GetPipelineResponse = {
  pipeline: {
    name: 'podinfo-02',
    namespace: 'flux-system',
    appRef: {
      apiVersion: 'helm.toolkit.fluxcd.io/v2beta1',
      kind: 'HelmRelease',
      name: 'podinfo',
    },
    environments: [
      {
        name: 'dev',
        promotion: { manual: true, strategy: { notification: {} } },
        targets: [
          {
            namespace: 'podinfo-02-dev',
            clusterRef: {
              kind: 'GitopsCluster',
              name: 'dev',
              namespace: '',
            },
          },
        ],
      },
      {
        name: 'test',
        promotion: { manual: true, strategy: { secretRef: { name: 'test' } } },
        targets: [
          {
            namespace: 'podinfo-02-qa',
            clusterRef: {
              kind: 'GitopsCluster',
              name: 'dev',
              namespace: '',
            },
          },
          {
            namespace: 'podinfo-02-perf',
            clusterRef: {
              kind: 'GitopsCluster',
              name: 'dev',
              namespace: '',
            },
          },
        ],
      },
      {
        name: 'prod',
        promotion: {
          manual: true,
          strategy: {
            pullRequest: {
              type: 'github',
              url: 'https://gitlab.com/weaveworks/cool-project',
              branch: 'main',
            },
          },
        },
        targets: [
          {
            namespace: 'podinfo-02-prod',
            clusterRef: {
              kind: 'GitopsCluster',
              name: 'prod',
              namespace: '',
            },
          },
        ],
      },
    ],
    targets: [],
    status: {
      environments: {
        dev: {
          waitingStatus: { revision: '2.0.0' },
          targetsStatuses: [
            {
              clusterRef: {
                kind: 'GitopsCluster',
                name: 'dev',
                namespace: 'flux-system',
              },
              namespace: 'podinfo-02-dev',
              workloads: [
                {
                  kind: 'HelmRelease',
                  name: 'podinfo',
                  version: '6.0.0',
                  lastAppliedRevision: '6.0.0.0',
                  conditions: [
                    {
                      type: 'Ready',
                      status: 'True',
                      reason: 'ReconciliationSucceeded',
                      message: 'Release reconciliation succeeded',
                      timestamp: '2022-12-07T15:06:00Z',
                    },
                    {
                      type: 'Released',
                      status: 'True',
                      reason: 'UpgradeSucceeded',
                      message: 'Helm upgrade succeeded',
                      timestamp: '2022-12-07T15:06:00Z',
                    },
                  ],
                  suspended: false,
                },
              ],
            },
          ],
        },
        prod: {
          targetsStatuses: [
            {
              clusterRef: {
                kind: 'GitopsCluster',
                name: 'prod',
              },
              namespace: 'podinfo-02-prod',
              workloads: [
                {
                  kind: 'HelmRelease',
                  name: 'podinfo',
                  version: '6.0.1',
                  lastAppliedRevision: '6.0.1.1',
                  conditions: [
                    {
                      type: 'Ready',
                      status: 'True',
                      reason: 'ReconciliationSucceeded',
                      message: 'Release reconciliation succeeded',
                      timestamp: '2022-09-20T09:06:30Z',
                    },
                    {
                      type: 'Released',
                      status: 'True',
                      reason: 'UpgradeSucceeded',
                      message: 'Helm upgrade succeeded',
                      timestamp: '2022-09-20T09:06:30Z',
                    },
                  ],
                  suspended: false,
                },
              ],
            },
          ],
        },
        test: {
          targetsStatuses: [
            {
              clusterRef: {
                kind: 'GitopsCluster',
                name: 'dev',
                namespace: 'flux-system',
              },
              namespace: 'podinfo-02-qa',
              workloads: [
                {
                  kind: 'HelmRelease',
                  name: 'podinfo',
                  version: '6.0.2',
                  lastAppliedRevision: '6.0.2.2',
                  conditions: [
                    {
                      type: 'Ready',
                      status: 'True',
                      reason: 'ReconciliationSucceeded',
                      message: 'Release reconciliation succeeded',
                      timestamp: '2022-09-20T09:07:01Z',
                    },
                    {
                      type: 'Released',
                      status: 'True',
                      reason: 'InstallSucceeded',
                      message: 'Helm install succeeded',
                      timestamp: '2022-09-20T09:07:01Z',
                    },
                  ],
                  suspended: false,
                },
              ],
            },
            {
              clusterRef: {
                kind: 'GitopsCluster',
                name: 'dev',
                namespace: 'flux-system',
              },
              namespace: 'podinfo-02-perf',
              workloads: [
                {
                  kind: 'HelmRelease',
                  name: 'podinfo',
                  version: '6.0.3',
                  lastAppliedRevision: '6.0.3.3',
                  conditions: [
                    {
                      type: 'Ready',
                      status: 'True',
                      reason: 'ReconciliationSucceeded',
                      message: 'Release reconciliation succeeded',
                      timestamp: '2022-10-13T13:37:34Z',
                    },
                    {
                      type: 'Released',
                      status: 'True',
                      reason: 'UpgradeSucceeded',
                      message: 'Helm upgrade succeeded',
                      timestamp: '2022-10-13T13:37:34Z',
                    },
                  ],
                  suspended: false,
                },
              ],
            },
          ],
        },
      },
    },
    yaml: 'apiVersion: pipelines.weave.works/v1alpha1\nkind: Pipeline\nmetadata:\n  creationTimestamp: "2022-10-25T16:49:48Z"\n  generation: 4\n  labels:\n    kustomize.toolkit.fluxcd.io/name: pipelines\n    kustomize.toolkit.fluxcd.io/namespace: flux-system\n  managedFields:\n  - apiVersion: pipelines.weave.works/v1alpha1\n    fieldsType: FieldsV1\n    fieldsV1:\n      f:metadata:\n        f:labels:\n          f:kustomize.toolkit.fluxcd.io/name: {}\n          f:kustomize.toolkit.fluxcd.io/namespace: {}\n      f:spec:\n        f:appRef:\n          f:apiVersion: {}\n          f:kind: {}\n          f:name: {}\n        f:environments: {}\n        f:promotion:\n          f:notification: {}\n          f:secretRef:\n            f:name: {}\n    manager: kustomize-controller\n    operation: Apply\n    time: "2022-12-08T13:29:01Z"\n  - apiVersion: pipelines.weave.works/v1alpha1\n    fieldsType: FieldsV1\n    fieldsV1:\n      f:status:\n        f:conditions: {}\n        f:observedGeneration: {}\n    manager: pipeline-controller\n    operation: Update\n    subresource: status\n    time: "2022-10-31T14:30:45Z"\n  name: podinfo-02\n  namespace: flux-system\n  resourceVersion: "230322663"\n  uid: cb61bc7a-9012-4d72-bedd-288699fbfabb\nspec:\n  appRef:\n    apiVersion: helm.toolkit.fluxcd.io/v2beta1\n    kind: HelmRelease\n    name: podinfo\n  environments:\n  - name: dev\n    targets:\n    - clusterRef:\n        kind: GitopsCluster\n        name: dev\n        namespace: flux-system\n      namespace: podinfo-02-dev\n  - name: test\n    targets:\n    - clusterRef:\n        kind: GitopsCluster\n        name: dev\n        namespace: flux-system\n      namespace: podinfo-02-qa\n    - clusterRef:\n        kind: GitopsCluster\n        name: dev\n        namespace: flux-system\n      namespace: podinfo-02-perf\n  - name: prod\n    targets:\n    - clusterRef:\n        kind: GitopsCluster\n        name: prod\n        namespace: default\n      namespace: podinfo-02-prod\n  promotion:\n    notification: {}\nstatus:\n  conditions:\n  - lastTransitionTime: "2022-10-31T14:30:45Z"\n    message: All clusters checked\n    reason: ReconciliationSucceeded\n    status: "True"\n    type: Ready\n  observedGeneration: 4\n',
    type: 'Pipeline',
  },
};

describe('PipelineDetails', () => {
  let wrap: (el: JSX.Element) => JSX.Element;
  let api: PipelinesClientMock;
  let core: CoreClientMock;

  beforeEach(() => {
    api = new PipelinesClientMock();
    core = new CoreClientMock();
    wrap = withContext([
      ...defaultContexts(),
      [EnterpriseClientContext.Provider, { value: { pipelines: api } }],
      [CoreClientContextProvider, { api: core }],
    ]);
  });
  it('renders pipeline details', async () => {
    const params = res.pipeline;
    api.GetPipelineReturns = res;
    core.GetObjectReturns = { object: {} };

    await act(async () => {
      const c = wrap(
        <PipelineDetails
          name={params?.name || ''}
          namespace={params?.namespace || ''}
        />,
      );
      render(c);
    });

    // Breadcrumbs
    const title = screen.getByTestId('link-Pipelines').textContent;
    expect(title).toEqual('Pipelines');
    expect(await screen.findByText('podinfo-02')).toBeTruthy();
    //3 envs
    expect(await screen.findByText('dev')).toBeInTheDocument();
    expect(await screen.findByText('test')).toBeInTheDocument();
    expect(await screen.findByText('prod')).toBeInTheDocument();
    //3 targets with applied and speicified versions
    let i = 0;
    while (i <= 3) {
      expect(
        await screen.findByText(`SPECIFIED VERSION: v6.0.${i}`),
      ).toBeInTheDocument();
      expect(
        await screen.findByText(`LAST APPLIED VERSION: v6.0.${i}.${i}`),
      ).toBeInTheDocument();
      i++;
    }
  });

  describe('renders promotion strategy', () => {
    it('pull request', async () => {
      const params = res.pipeline;
      api.GetPipelineReturns = res;
      core.GetObjectReturns = { object: {} };

      await act(async () => {
        const c = wrap(
          <PipelineDetails
            name={params?.name || ''}
            namespace={params?.namespace || ''}
          />,
        );
        render(c);
      });
      expect(screen.getByText('Pull Request')).toBeInTheDocument();
      expect(
        screen.getByText('https://gitlab.com/weaveworks/cool-project'),
      ).toBeInTheDocument();
      expect(screen.getByText('main')).toBeInTheDocument();
      expect(screen.getByText('Secret Ref')).toBeInTheDocument();
      expect(screen.getByText('Notification')).toBeInTheDocument();
    });
  });
  it('handles visibility of promotion button', async () => {
    const params = res.pipeline;
    const manual: Pipeline = {
      ...res.pipeline,
      promotion: {
        manual: true,
      },
    };
    api.GetPipelineReturns = { ...res, pipeline: manual };
    core.GetObjectReturns = { object: {} };

    await act(async () => {
      const c = wrap(
        <PipelineDetails
          name={params?.name || ''}
          namespace={params?.namespace || ''}
        />,
      );
      render(c);
    });
    const buttons = screen.getAllByRole('button');

    const filteredButtons = buttons.filter(e =>
      e.innerHTML.includes('Approve Promotion'),
    );
    expect(filteredButtons.length).toEqual(2);
    const devButton = filteredButtons.filter(
      e => !e.className.includes('Mui-disabled'),
    );
    expect(devButton.length).toEqual(1);
  });
});
