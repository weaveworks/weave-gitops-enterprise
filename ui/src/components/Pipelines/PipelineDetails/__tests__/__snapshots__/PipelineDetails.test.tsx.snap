// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PipelineDetails renders pipeline Yaml 1`] = `
"1apiVersion: pipelines.weave.works/v1alpha1
2kind: Pipeline
3metadata:
4  creationTimestamp: \\"2022-10-25T16:49:48Z\\"
5  generation: 4
6  labels:
7    kustomize.toolkit.fluxcd.io/name: pipelines
8    kustomize.toolkit.fluxcd.io/namespace: flux-system
9  managedFields:
10  - apiVersion: pipelines.weave.works/v1alpha1
11    fieldsType: FieldsV1
12    fieldsV1:
13      f:metadata:
14        f:labels:
15          f:kustomize.toolkit.fluxcd.io/name: {}
16          f:kustomize.toolkit.fluxcd.io/namespace: {}
17      f:spec:
18        f:appRef:
19          f:apiVersion: {}
20          f:kind: {}
21          f:name: {}
22        f:environments: {}
23        f:promotion:
24          f:notification: {}
25          f:secretRef:
26            f:name: {}
27    manager: kustomize-controller
28    operation: Apply
29    time: \\"2022-12-08T13:29:01Z\\"
30  - apiVersion: pipelines.weave.works/v1alpha1
31    fieldsType: FieldsV1
32    fieldsV1:
33      f:status:
34        f:conditions: {}
35        f:observedGeneration: {}
36    manager: pipeline-controller
37    operation: Update
38    subresource: status
39    time: \\"2022-10-31T14:30:45Z\\"
40  name: podinfo-02
41  namespace: flux-system
42  resourceVersion: \\"230322663\\"
43  uid: cb61bc7a-9012-4d72-bedd-288699fbfabb
44spec:
45  appRef:
46    apiVersion: helm.toolkit.fluxcd.io/v2beta1
47    kind: HelmRelease
48    name: podinfo
49  environments:
50  - name: dev
51    targets:
52    - clusterRef:
53        kind: GitopsCluster
54        name: dev
55        namespace: flux-system
56      namespace: podinfo-02-dev
57  - name: test
58    targets:
59    - clusterRef:
60        kind: GitopsCluster
61        name: dev
62        namespace: flux-system
63      namespace: podinfo-02-qa
64    - clusterRef:
65        kind: GitopsCluster
66        name: dev
67        namespace: flux-system
68      namespace: podinfo-02-perf
69  - name: prod
70    targets:
71    - clusterRef:
72        kind: GitopsCluster
73        name: prod
74        namespace: default
75      namespace: podinfo-02-prod
76  promotion:
77    notification: {}
78status:
79  conditions:
80  - lastTransitionTime: \\"2022-10-31T14:30:45Z\\"
81    message: All clusters checked
82    reason: ReconciliationSucceeded
83    status: \\"True\\"
84    type: Ready
85  observedGeneration: 4
86"
`;
